docs: add blog: Utilizing Your Private Git Repository Context for Enhanced Code Completion (#1705)
TabbyML/tabby: website/blog/2024-03-21-using-your-private-git-repository-context-for-completion/index.md

---
authors: [icycodes]
tags: [deployment, repository context]
---

# Connect Private GitHub Repository to Tabby

A few months back, we published a blog [Repository context for LLM assisted code completion](https://tabby.tabbyml.com/blog/2023/10/16/repository-context-for-code-completion), introducing the `Repository Context` feature in Tabby.

> In this blog, I will guide you through the process of setting up a Tabby server configured with a private Git repositories context.

Reviewer: ```suggestion
In this blog, I will guide you through the process of setting up a Tabby server configured with **a private Git repositories** context.
```

> In this blog, I will guide you through the process of setting up a Tabby server configured with **a private Git repositories** context.

Reviewer: Consider adding something suggesting that many users have adopted this to incorporate repo-level knowledge into Tabby (to highlight the wide adoption). In order to make this integration process easier and more smooth, we are introducing how to set up Tabby server with private Git repos in this blog.

## Generating a Personal Access Token

> In order to provide the Tabby server with access to your private Git repositories, it is essential to create a Personal Access Token (PAT) specific to your Git provider. The following steps outline the process with GitHub as a reference:

Reviewer: ```suggestion
In order to provide the Tabby server with access to your private Git repositories, it is essential to create a ##Personal Access Token (PAT)## specific to your Git provider. The following steps outline the process with GitHub as a reference:
```

1. Visit [GitHub Personal Access Tokens Settings](https://github.com/settings/tokens?type=beta) and select `Generate new token`.
   ![GitHub PAT Generate New Token](./github-pat-generate-new-token.png)
2. Enter the `Token name`, specify an `Expiration` date, an optional `Description`, and select the repositories you wish to grant access to.
   ![GitHub PAT Filling Info](./github-pat-filling-info.png)
3. Within the `Permissions` section, ensure that `Contents` is configured for `Read-only` access.
   ![GitHub PAT Contents Access](./github-pat-contents-access.png)
4. Click `Generate token` to generate the new PAT. Remember to make a copy of the PAT before closing the webpage.
   ![GitHub PAT Generate Token](./github-pat-generate-token.png)

> For additional information, please consult the documentation on [Managing your personal access tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens).

Reviewer: ```suggestion
For additional information, please refer to the documentation on [Managing your personal access tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens).
```

**Note**: For users of GitLab, guidance on creating a personal access token can be found in the documentation [Personal access tokens - GitLab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token).

## Configuration

To configure the Tabby server with your private Git repositories, you need to provide the required settings in a YAML file. Create and edit a configuration file located at `~/.tabby/config.yaml`:

```yaml
## Add the private repository
[[repositories]]
name = "my_private_project"
git_url = "https://<PAT>@github.com/icycodes/my_private_project.git"

## More repositories can be added like this
[[repositories]]
name = "another_project"
git_url = "https://<PAT>@github.com/icycodes/another_project.git"
```

For more detailed about the configuration file, you can refer to the [configuration documentation](https://tabby.tabbyml.com/docs/configuration).

**Note:** The URL format for GitLab repositories may vary, you can check the [official documentation](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#clone-repository-using-personal-access-token) for specific guidelines.

## Building the Index

> 

Reviewer: Add something here to briefly talk about how we are building the index - for example, tree-sitter is used to parse your code and extract critical code components for index building. The index is for fast retrieval (related code snippets) during code completion time to enhance code generation context. Something like this would be helpful.

Author: Updated in [08ade79](https://github.com/TabbyML/tabby/pull/1705/commits/08ade79e47c2db2bf37431b711e34765a0ac3e97)

2024-03-21T16:16:47.189632Z  INFO tabby::serve: crates/tabby/src/serve.rs:118: Starting server, this might take a few minutes...
2024-03-21T16:16:47.190764Z  INFO tabby::services::code: crates/tabby/src/services/code.rs:53: Index is ready, enabling server...
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes
2024-03-21T16:16:52.464116Z  INFO tabby::routes: crates/tabby/src/routes/mod.rs:35: Listening at 0.0.0.0:8080
```

Notably, the line `Index is ready, enabling server...` signifies that the server has been successfully launched with the constructed index.

## Verifying Indexing Results

To confirm that the code completion is effectively utilizing the built index, you can employ the code search feature to validate the indexing process:

1. Access the Swagger UI page: `http://localhost:8080/swagger-ui/#/v1beta/search`.
2. Click on the `Try it out` button, and input the query parameter `q`, such as `get`.
3. Click the `Execute` button to trigger the search and see if there are any indexed symbols.

> 

Reviewer: Add a screenshot for the code search experience?

Author: Updated in [01e4d76](https://github.com/TabbyML/tabby/pull/1705/commits/01e4d7675bc7af89b1b088f8bddff944df184412)

Alternatively, if you have utilized the code completion with the constructed index, you can examine the server log located in `~/.tabby/events` to inspect how the prompt is enhanced during code completion.

## Additional Notes

> Starting from version v0.9, Tabby Enterprise Edition offers a web UI to manage your git repository contexts. Additionally, a scheduler job management system has been integrated, streamlining the process of monitoring scheduler job statuses. With these enhancements, you can save a lot of effort in maintaining yaml config files and docker compose configurations. Furthermore, users can easily monitor visualized indexing results through the built-in code browser.

Reviewer 2: ```suggestion
Starting from version v0.9, Tabby offers a web UI to manage your git repository contexts. Additionally, a scheduler job management system has been integrated, streamlining the process of monitoring scheduler job statuses. With these enhancements, you can save a lot of effort in maintaining yaml config files and docker compose configurations. Furthermore, users can easily monitor visualized indexing results through the built-in code browser.
```

In the upcoming v0.10, a new feature will be introduced that enables a direct connection to GitHub, simplifying and securing your access to private GitHub repositories.

> For further details and guidance, please refer to our enterprise edition documentation.

Reviewer 2: ```suggestion
For further details and guidance, please refer to [administration documents](https://tabby.tabbyml.com/docs/administration/).
```

---
authors: [icycodes]
tags: [deployment, repository context]
---

> # Utilizing Your Private Git Repository Context for Enhanced Code Completion

Reviewer 2: ```suggestion
# Connect Private Github Repository to Tabby
```

A few months back, we published a blog [Repository context for LLM assisted code completion](https://tabby.tabbyml.com/blog/2023/10/16/repository-context-for-code-completion), introducing the `Repository Context` feature in Tabby. 

In this blog, I will guide you through the process of setting up a Tabby server configured with a private Git repositories context.

> ## Generating a Personal Access Token for Your Private Git Repository

Reviewer 2: ```suggestion
## Generating a Personal Access Token
```

  ![GitHub PAT Generate New Token](./github-pat-generate-new-token.png)
2. Enter the `Token name`, specify an `Expiration` date, an optional `Description`, and select the repositories you wish to grant access to.
  ![GitHub PAT Filling Info](./github-pat-filling-info.png)
3. Within the `Permissions` section, ensure that `Contents` is configured for `Read-only` access.
  ![GitHub PAT Contents Access](./github-pat-contents-access.png)
4. Click `Generate token` to generate the new PAT. Remember to make a copy of the PAT before closing the webpage.
  ![GitHub PAT Generate Token](./github-pat-generate-token.png)

For additional information, please consult the documentation on [Managing your personal access tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens).

**Note**: For users of GitLab, guidance on creating a personal access token can be found in the documentation [Personal access tokens - GitLab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token).

> ## Creating a Configuration File for Tabby Server

Reviewer 2: use `--webserver` based configuration (as git repository is part of community tier)

> ## Creating a Configuration File for Tabby Server

Reviewer 2: ```suggestion
## Configuration
```

To configure the Tabby server with your private Git repositories, you need to provide the required settings in a YAML file. Create and edit a configuration file located at `~/.tabby/config.yaml`:

```yaml
## Add the private repository
[[repositories]]
name = "my_private_project"
git_url = "https://<PAT>@github.com/icycodes/my_private_project.git"

## More repositories can be added like this
[[repositories]]
name = "another_project"
git_url = "https://<PAT>@github.com/icycodes/another_project.git"
```

For more detailed about the configuration file, you can refer to the [configuration documentation](https://tabby.tabbyml.com/docs/configuration).

**Note:** The URL format for GitLab repositories may vary, you can check the [official documentation](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#clone-repository-using-personal-access-token) for specific guidelines.

> ## Building the Index and Starting the Server

Reviewer 2: use `code browser for validation`

> ## Building the Index and Starting the Server

Reviewer 2: ```suggestion
## Building the index
```

```

Subsequently, launch the server using the following command:

```bash
docker run -it --gpus all -p 8080:8080 -v $HOME/.tabby:/data tabbyml/tabby serve --model StarCoder-1B --device cuda
```

The expected output upon successful initiation of the server should like this:

```console
icy@Icys-Ubuntu:~$ docker run -it --gpus all -p 8080:8080 -v $HOME/.tabby:/data tabbyml/tabby serve --model StarCoder-1B --device cuda
2024-03-21T16:16:47.189632Z  INFO tabby::serve: crates/tabby/src/serve.rs:118: Starting server, this might take a few minutes...
2024-03-21T16:16:47.190764Z  INFO tabby::services::code: crates/tabby/src/services/code.rs:53: Index is ready, enabling server...    
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes
2024-03-21T16:16:52.464116Z  INFO tabby::routes: crates/tabby/src/routes/mod.rs:35: Listening at 0.0.0.0:8080
```

Notably, the line `Index is ready, enabling server...` signifies that the server has been successfully launched with the constructed index.

> ### Automated Scheduler Execution (Optional)

Reviewer 2: remove this section.

> ### Automated Scheduler Execution (Optional)
> 
> By excluding the `--now` flag from the `scheduler` command, you can configure the scheduler to run automatically at intervals, typically every 10 minutes by default.
> 
> To integrate the scheduler with the server, a Docker Compose file can be created as shown below:
> 
> ```yaml
> version: '3.5'
> name: tabby
> services:
>   serve:
>     restart: always
>     image: tabbyml/tabby
>     command: serve --model StarCoder-1B --device cuda
>     ports:
>       - "8080:8080"
>     volumes:
>       - "~/.tabby:/data"
>     deploy:
>       resources:
>         reservations:
>           devices:
>             - driver: nvidia
>               count: 1
>               capabilities: [gpu]
>   scheduler:
>     restart: always
>     image: tabbyml/tabby
>     entrypoint: /opt/tabby/bin/tabby-cpu
>     command: scheduler
>     volumes:
>       - "~/.tabby:/data"
> ```

Reviewer 2: ```suggestion
```

> ## Verifying Indexing Results

Reviewer 2: use code browser to validate.

To confirm that the code completion is effectively utilizing the built index, you can employ the code search feature to validate the indexing process:

1. Access the Swagger UI page: `http://localhost:8080/swagger-ui/#/v1beta/search`.
2. Click on the `Try it out` button, and input the query parameter `q`, such as `get`.
3. Click the `Execute` button to trigger the search and see if there are any indexed symbols.

Alternatively, if you have utilized the code completion with the constructed index, you can examine the server log located in `~/.tabby/events` to inspect how the prompt is enhanced during code completion.

> ## Additional Information on Tabby Enterprise Edition

Reviewer 2: ```suggestion
## Additional Notes
```

> Starting from version v0.9.0, Tabby Enterprise Edition offers a web UI to manage your git repository contexts. Additionally, a scheduler job management system has been integrated, streamlining the process of monitoring scheduler job statuses. With these enhancements, you can save a lot of effort in maintaining yaml config files and docker compose configurations. Furthermore, users can easily monitor visualized indexing results through the built-in code browser.

Reviewer 2: mention that in 0.10.0 we'll support connect directly with github.

Author: Updated in 8b5e1e04abfb2ee6ca6d06c523c90059383de006

> Starting from version v0.9.0, Tabby Enterprise Edition offers a web UI to manage your git repository contexts. Additionally, a scheduler job management system has been integrated, streamlining the process of monitoring scheduler job statuses. With these enhancements, you can save a lot of effort in maintaining yaml config files and docker compose configurations. Furthermore, users can easily monitor visualized indexing results through the built-in code browser.

Reviewer 2: ```suggestion
Starting from version v0.9, Tabby Enterprise Edition offers a web UI to manage your git repository contexts. Additionally, a scheduler job management system has been integrated, streamlining the process of monitoring scheduler job statuses. With these enhancements, you can save a lot of effort in maintaining yaml config files and docker compose configurations. Furthermore, users can easily monitor visualized indexing results through the built-in code browser.
```

Use semver minor
